### PR TITLE
fall back to system ruff executable without executable setting

### PR DESCRIPTION
**Pull request by TheJJ**
otherwise I have to either configure the path in the project's pyproject.toml or unnecessarily in the lsp configuration :)

### CODE REVIEW

**Comment by TheJJ** [comment]
Hi! - no worries :)
Yes, `python3 -m ruff` does not work for me, since the Gentoo `dev-util/ruff` package doesn't install the Python module
Apart from shell completions and documentation, I only have `/usr/bin/ruff`.

I think it's alright to use `shutil.which` to search in the `PATH` for the executable, since I don't see a useful way to hard-code the executable path in the in-project `pyproject.toml`, and hard-coding it in the lsp configuration doesn't seem right when it's just available in the PATH.

What do you think?

**Comment by jhossbach** [comment]
Instead of using importlib, I moved finding the correct executable to a new function which is now cached, so it should only run once with the given executable.
Can you try to see if that works?

File: pylsp_ruff/plugin.py [hunk 1]
```diff
@@ -487,6 +487,24 @@ def run_ruff_format(
...
+            Popen(cmd, stdin=PIPE, stdout=PIPE, stderr=PIPE).communicate()
+        except Exception:
+            log.error(f"Can't execute ruff with given executable '{executable}'.")
+        finally:
+            return cmd
+    else:
+        cmd = [sys.executable, "-m", "ruff"]
+        (_, stderr) = Popen(cmd, stdin=PIPE, stdout=PIPE, stderr=PIPE).communicate()
+        if "No module named ruff" in stderr.decode():
+            cmd = [shutil.which("ruff")]
```
Comment (TheJJ) [inline pylsp_ruff/plugin.py:504]: this seems rather hacky, especially just looking at the output, which i assume is not stable for that error message :)

you could instead try to import the ruff module (try: import ruff except ImportError), or as mentioned before check if it exists as entry point.

**Comment by jhossbach** [comment]
This should work now, please check if it does.

**Review by TheJJ — COMMENTED** [review]
from what I understand now, what you whish is:
- if `executable` is set and exists, use it
- otherwise, try the ruff module
- otherwise, use the system executable

I'd propose to implement is this way:
- if `executable` is set, use it, and crash if it doesn't exist. silent swiches of the used program are unexpected and hard to debug.
- otherwise, check once if the ruff module exists and use it
- otherwise, use the system executable found by `which`.

maybe some way to prefer the executable would be better, because then we could bypass invoking a new python interpreter for every ruff invocation eats a bit of latency. or calling the python module in the same interpreter.

**Comment by TheJJ** [comment]
I've implemented an improved version, which uses the configuration value without fallback, or it tries the python module and then system executable next.

one could optimize this to eliminate the extra python interpreter invocation if using the module (for latency improvement) by re-using the current interpreter instead.

**Comment by jhossbach** [comment]
Looks really nice, but I am not sure the ImportError is propagated the way you probably want it to, e.g. when testing with neovim the error does not reach the editor.
I think there is a `workspace.show_message` method which could work here (although I haven't tested it).

**Comment by TheJJ** [comment]
which import error, if no executable/module is found?

if there's a better way than crashing with an exception, we should surely use it :)

**Comment by jhossbach** [comment]
Similarly to `log.error`, the RunTimeError will only be seen in the pylsps log file.
I just tried to try except the Error for a dummy executable and report it to the editor with `workspace.show_message(...)` but so far this did not work for me.
Unfortunately I don't have the time to look into it until next week.

**Comment by TheJJ** [comment]
maybe we could merge this for now since it's not making it worse than before when no ruff was found :)
then come up with sane error propagation to notify the user properly.

**Comment by jhossbach** [comment]
I took the liberty of removing the RuntimeErrors for now, otherwise thanks a lot for the PR!

**Comment by TheJJ** [comment]
great :)
thanks for the tool!

**Pull request merged**